examples/run_nlp_trainer.py

```diff
@@ -0,0 +1,439 @@
+    cuda_envs = mp.Queue(queue_size)
+    cuda_envs_list = range(queue_size)
+    for e in cuda_envs_list:
+        cuda_envs.put(e)
+
+    return cuda_envs
+
+
+def init_cuda_env(cuda_envs: mp.Queue) -> None:
+    """Set up a single GPU CUDA environment on initialization of a mp process pool."""
+    env = cuda_envs.get()
+    torch.cuda.set_device(env)
+
+    # fake a single node world
+    os.environ['WORLD_SIZE'] = '1'
+    os.environ['LOCAL_WORLD_SIZE'] = '1'
+
+
+def get_all_s3_checkpoints(bucket_name: str, folder_name: str) -> List[str]:
+    """List all the checkpoints in a given S3 bucket and folder."""
+    # pulls AWS credentials from environment variables, assumes that aws_session_token is defined
+    session = boto3.Session()
+
+    s3 = session.resource('s3')
+    bucket = s3.Bucket(bucket_name)
+    ckpt_names = []
+    for obj in bucket.objects.all():
+        if folder_name in obj.key:
+            obj_name = obj.key
+            ckpt_names.append(obj_name)
+
+    return ckpt_names
+
+
+def output_metrics(glue_metrics: GlueState) -> None:
+    """Consolidate and prettify metrics."""
+    tasks = glue_metrics.task_names
+    large_tasks = ['mnli', 'qnli', 'qqp', 'sst2']
+    # init table headers
+    headers = ['Checkpoint']
+    headers.extend([f'{task.upper()}' for task in sorted(tasks)])
+    headers.extend(['GLUE-Large', 'GLUE-All'])
+    tb = [headers]
+
+    # fill table
+    for ckpt in glue_metrics.ckpt_to_tasks.keys():
+        output_line = [ckpt]
+        glue_all = 0
+        glue_large = 0
+        # Per task score
+        for task in sorted(glue_metrics.task_names):
+            task_metric = 0 if not glue_metrics.ckpt_to_tasks[ckpt].task_to_avg_metric[
+                task] else glue_metrics.ckpt_to_tasks[ckpt].task_to_avg_metric[task]
+            output_line.append('{:.4f}'.format(task_metric))
+            glue_all += task_metric
+            if task in large_tasks:
+                glue_large += task_metric
+        # GLUE Large and GLUE All
+        output_line.append('{:.4f}'.format(glue_large / len(large_tasks)))
+        output_line.append('{:.4f}'.format(glue_all / len(tasks)))
+        tb.append(output_line)
+
+    print(tabulate(tb, headers='firstrow'))
+
+
+def log_metrics(ckpt_filename: str, glue_metrics: GlueState, f_metric: Future) -> None:
+    """Callback function for metric collection."""
+    metric = f_metric.result()
+
+    if ckpt_filename not in glue_metrics.ckpt_to_tasks.keys():
+        glue_metrics.ckpt_to_tasks[ckpt_filename] = GLUEMetricsLogger(glue_metrics.task_names)
+
+    task = list(metric.keys())[0]
+    formatted_task = task.split('_')[1]  # remove "glue_" prefix
+    for key in metric.keys():  # handle case where mnli has glue_mnli and glue_mnli_mismatched
+        for m in metric[key]:
+            if not glue_metrics.ckpt_to_tasks[ckpt_filename].task_to_avg_metric[formatted_task]:
```

Comment:
Would it work to put back the helper methods that were previously defined in the GLUEState class? I think you mentioned previously that because it's only used in one or two places that I should just directly put it in the code.

Reply:
Would it make sense to put back the helper methods in the GlueState class? One of your previous comments suggested to just discard the helper methods and put it directly in the code so it seems like I'm going back and forth on this.

Reply:
Ah, by a variable, I mean something defined inline here, like:

```python
tasks = glue_metrics.ckpt_to_tasks[ckpt_filename]
metric = tasks.task_to_avg_metric[formatted_task]
```